feat: curio: add miner init

LexLuthr opened the pull request:
## Related Issues
<!-- Link issues that this PR might resolve/fix. If an issue doesn't exist, include a brief motivation for the change being made -->

fixes https://github.com/filecoin-project/lotus/issues/11758

## Proposed Changes
<!-- A clear list of the changes being made -->
This PR adds a miner creation option to Curio guided-setup.

```
Defaulting to English. Please reach out to the Curio team if you would like to have additional language support.
✔ Create a new miner
                                                                
 ┌────────────────────────────────────────────────────────────┐ 
 │This interactive tool creates a new miner actor and creates │ 
 │the basic configuration layer for it.                       │ 
 └────────────────────────────────────────────────────────────┘ 
                                                                
This process is partially idempotent. Once a new miner actor has been created and subsequent steps fail, the user need to run 'curio config new-cluster < miner ID >' to finish the configuration.
                                                                                                                                                                                                  
✔ Port: 5433
Enter the Yugabyte database port: 5435
✔ Continue to connect and update schema.
✔ Step Complete: Pre-initialization steps complete
                                                
Initializing a new miner actor.
✔ Owner Address: <empty>
Enter the owner address: t3rgs3idbhbpk2myj5cdvcxxuwznhxefo6vdcr4n6ohaa5fepsxgcwax4oqazp2ypvp5ay2qhcjbr6uykycfua
✔ Enter worker address: █3rgs3idbhbpk2myj5cdvcxxuwznhxefo6vdcr4n6ohaa5fepsxgcwax4oqazp2ypvp5ay2qhcjbr6uykycfuaEnter worker address: t3rgs3idbhbpk2myj5cdvcxxuwznhxefo6vdcr4n6ohaa5fepsxgcwax4oqazp2ypvp5ay2qhcjbr6uykycfua
✔ Sender Address: <empty>
Enter sender address: t3rgs3idbhbpk2myj5cdvcxxuwznhxefo6vdcr4n6ohaa5fepsxgcwax4oqazp2ypvp5ay2qhcjbr6uykycfua
✔ Sector Size: 0
Enter the sector size: 8mib
✔ Confidence epochs: 0
Enter the confidence: 5
Pushed CreateMiner message: bafy2bzacedlxo6syqno5op5ryyqk4s7ivkbpvcs56ro6pyq42tixtccdsclwm
Waiting for confirmation
New miners address is: t01012 (t2uchxg6gfvpbeumr46mt3rlaedgitgqhheuu2gmi)
✔ Step Complete: Miner t01012 created successfully
                                                
✔ Step Complete: New Curio configuration layer 't01012' created
                                                             
Documentation: 
The 'base' layer stores common configuration. All curio instances can include it in their --layers argument.
You can add other layers for per-machine configuration changes.
Filecoin Slack channels: #fil-curio-help and #fil-curio-dev
Start multiple Curio instances with the 'post' layer to redundancy.
One database can serve multiple miner IDs: Run a migration for each lotus-miner.
The Curio team wants to improve the software you use. Tell the team you're using `curio`.
✔ Nothing.
✔ Step Complete: New Miner initialization complete.
                                                 
Try the web interface with --layers=gui for further guided improvements.
```

## Additional Info
<!-- Callouts, links to documentation, and etc -->

## Checklist

Before you mark the PR ready for review, please make sure that:

- [x] Commits have a clear commit message.
- [x] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] If the PR affects users (e.g., new feature, bug fix, system requirements change), update the CHANGELOG.md and add details to the UNRELEASED section.
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

magik6k commented:
IMO miner initialization doesn't belong in the new-cluster command.

I'd put the basic miner initialize command in `sptool` e.g. as `sptool actor create`, and then separately make it possible to initialize a miner actor in `curio guided-setup` interactive flow.

snadrus commented:
Please fix my poor doc from
```
Start multiple Curio instances with the 'post' layer to redundancy.
```
to:
```
Increase reliability using redundancy: start multiple machines with at-least the post layer: `curio run --layers=post`
```

LexLuthr commented:
> Please fix my poor doc from
> 
> ```
> Start multiple Curio instances with the 'post' layer to redundancy.
> ```
> 
> to:
> 
> ```
> Increase reliability using redundancy: start multiple machines with at-least the post layer: `curio run --layers=post`
> ```

Done

snadrus commented on cmd/curio/guidedsetup/guidedsetup.go:
```diff
@@ -164,6 +181,23 @@ func SetupLanguage() (func(key message.Reference, a ...interface{}) string, func
 
+func newOrMigrate(d *MigrationData) {
+	i, _, err := (&promptui.Select{
+		Label: d.T("Choose if you with to create a new miner or migrate from existing Lotus-Miner"),
```
```suggestion
		Label: d.T("I want to:"),
```

snadrus commented on cmd/curio/guidedsetup/guidedsetup.go:
```diff
@@ -164,6 +181,23 @@ func SetupLanguage() (func(key message.Reference, a ...interface{}) string, func
+		Label: d.T("Choose if you with to create a new miner or migrate from existing Lotus-Miner"),
+		Items: []string{
+			d.T("Migrate from existing Lotus-Miner"),
+			d.T("Create a new miner")},
```
```suggestion
			d.T("Create a new miner with info from `%s` ", "sptool create miner")},
```
If that's where it will be.

LexLuthr replied:
We have a separate command for creating config. We use that in case guided-setup fails. We don't need `sptool create-miner` for the guided-setup.

LexLuthr commented:
@snadrus I have refactored guided-setup to add the new miner to base layer. I have also updated the `curio config new-cluster` command to do the same.

snadrus commented on cmd/curio/guidedsetup/guidedsetup.go:
```diff
@@ -783,16 +785,46 @@ func stepNewMinerConfig(d *MigrationData) {
+		d.say(notice, "Please do not run guided-setup again as miner creation is not idempotent. You need to run 'curio config new-cluster %s' to finish the configuration", d.MinerID.String())
+		os.Exit(1)
+	}
+	EnsureEmptyArrays(baseCfg)
```
no good. If they're multi-miner, this would zero-out their first miner ID.
Lets preserve pre-existing address for the multi-miner case.

LexLuthr replied:
Yes, I noticed that. I will be fix all the current issues and test it again in devnet.

snadrus commented on cmd/curio/guidedsetup/shared.go:
```diff
@@ -242,7 +242,7 @@ func getDBSettings(smCfg config.StorageMiner) string {
+func EnsureEmptyArrays(cfg *config.CurioConfig) {
 	if cfg.Addresses == nil {
 		cfg.Addresses = []config.CurioAddresses{}
 	} else {
```
what use-case does this solve?
Why erase good values?

snadrus commented:
Please ensure this workflow makes it into at-least a "Curio local net" setup document.

LexLuthr commented:
I have added AI in docs PR for Curio devnet documentation.

snadrus commented on node/config/load.go:
```diff
@@ -395,8 +398,32 @@ func ConfigUpdate(cfgCur, cfgDef interface{}, opts ...UpdateCfgOpt) ([]byte, err
+					return true
+				}
+				// both []string
+				if reflect.TypeOf(x) == reflect.TypeOf(y) {
```
They both should always be []string even with nil and {}. (Though I don't know if it's guaranteed to be the same pointer). You just be able to presume you have a []string
but if you want to verify it, do this instead:

tx, ty := reflect.TypeOf(x), reflect.TypeOf(y)
if tx.Kind() == reflect.Array 
&& ty.Kind() == reflect.Array 
&& tx.Elem().Kind() == reflect.String 
&& ty.Elem().Kind() == reflect.String

All you really need to verify if that the contents match:

```return strings.Join(x, "\n") == strings.Join(y,"\n")```    // possibly sorted first

LexLuthr replied:
I went with the type match. If both are Slice and elements are strings then we have a match.
Pushed the commit.

snadrus commented on node/config/load.go:
```diff
@@ -395,8 +398,32 @@ func ConfigUpdate(cfgCur, cfgDef interface{}, opts ...UpdateCfgOpt) ([]byte, err
+					return true
+				}
+				// one of them is nil
+				if reflect.TypeOf(x) == nil && reflect.TypeOf(x) != reflect.TypeOf(y) {
```
reflect.TypeOf( anything) != nil 
so these 2 cases (Line 414 & 418) can never happen

LexLuthr replied:
My bad! Nil is just a Pointer. Not sure what I was thinking there.
